xmodule/services.py

```diff
@@ -308,3 +312,81 @@ def _handle_deprecated_progress_event(self, block, event):
         # in order to avoid duplicate work and possibly conflicting semantics.
         if not getattr(block, 'has_custom_completion', False):
             self.completion_service.submit_completion(block.scope_ids.usage_id, 1.0)
+
+
+class ShowAnswerService(Service):
+    """
+    An XBlock Service that allows XModules to define if the answers are visible.
+    """
+    @classmethod
+    def answer_available(
+            cls, show_answer='', show_correctness='', past_due=False, attempts=0,
+            is_attempted=False, is_correct=False, required_attempts=0,
+            max_attempts=0, used_all_attempts=False, closed=False,
+            has_staff_access=False):
+        """
+        Returns whether correctness is available now, for the given attributes.
+        """
+        if not show_correctness:
+            # If correctness is being withheld, then don't show answers either.
+            return False
+        elif show_answer == ShowAnswer.NEVER:
+            return False
+        elif has_staff_access:
+            # This is after the 'never' check because course staff can see correctness
+            # unless the sequence/problem explicitly prevents it
+            return True
+        elif show_answer == ShowAnswer.ATTEMPTED:
+            return is_attempted or past_due
+        elif show_answer == ShowAnswer.ANSWERED:
+            # NOTE: this is slightly different from 'attempted' -- resetting the problems
+            # makes lcp.done False, but leaves attempts unchanged.
+            return is_correct
+        elif show_answer == ShowAnswer.CLOSED:
+            return closed
+        elif show_answer == ShowAnswer.FINISHED:
+            return closed or is_correct
+        elif show_answer == ShowAnswer.CORRECT_OR_PAST_DUE:
+            return is_correct or past_due
+        elif show_answer == ShowAnswer.PAST_DUE:
+            return past_due
+        elif show_answer == ShowAnswer.AFTER_SOME_NUMBER_OF_ATTEMPTS:
+            if max_attempts and required_attempts >= max_attempts:
+                required_attempts = max_attempts
+            return attempts >= required_attempts
+        elif show_answer == ShowAnswer.ALWAYS:
+            return True
+        elif show_answer == ShowAnswer.AFTER_ALL_ATTEMPTS:
+            return used_all_attempts
+        elif show_answer == ShowAnswer.AFTER_ALL_ATTEMPTS_OR_CORRECT:
+            return used_all_attempts or is_correct
+        elif show_answer == ShowAnswer.ATTEMPTED_NO_PAST_DUE:
+            return is_attempted
+        return False
+
+
+class ShowCorrectnessService(Service):
+    """
+    An XBlock Service that allows XModules to define whether correctness is currently hidden.
+
+    When correctness is hidden, this limits the user's access to the correct/incorrect flags, messages, problem scores,
+    and aggregate subsection and course grades.
+    """
+    @classmethod
+    def correctness_available(cls, show_correctness='', due_date=None, has_staff_access=False):
```

Comment:
We could eliminate the requirement of passing `show_correctness` and `due_date` to this method. [show_correctness](https://github.com/openedx/edx-platform/blob/f13263df4f0e9aff48fbbb7ca04583619317b947/xmodule/modulestore/inheritance.py#L102-L112) and [due](https://github.com/openedx/edx-platform/blob/f13263df4f0e9aff48fbbb7ca04583619317b947/xmodule/modulestore/inheritance.py#L46-L50)  are the attributes added to every XBlock through the [InheritanceMixin](https://github.com/openedx/edx-platform/blob/f13263df4f0e9aff48fbbb7ca04583619317b947/xmodule/modulestore/inheritance.py#L33).

To do this, we should add the following method:
```python
    def __init__(self, block=None, user_is_staff=False, **kwargs):
        super().__init__(**kwargs)
        self._user_is_staff = user_is_staff
        # This is needed because the `Service` class initialization expects the XBlock passed as an `xblock` keyword
        #  argument, but the `service` method from the `DescriptorSystem` passes a `block`.
        self._xblock = self.xblock() or block
```

Then, we would change the initialization in `block_render.py` to something like `'show_answer': partial(ShowAnswerService(), user_is_staff=user_is_staff),` and in `preview.py`: `'show_answer': partial(ShowAnswerService(), user_is_staff=True),`.

This way, we can access the XBlock directly in the service with `sef.xblock()`. It would make invoking these services easier. The [ReplaceURLService](https://github.com/openedx/edx-platform/blob/71fee4a4a044777e6725109181cda656f0b88dc9/common/djangoapps/static_replace/services.py) uses the same mechanism.

The `InheritanceMixin` also provides `max_attempts` and `showanswer`, so the `answer_available` method could be optimized in the same way.